README.md
<p align="center"><strong>Auto Install & Optimize LEMP Stack on Ubuntu (18.04, 20.04)</strong></p>
<p align="center"><strong>Author: Sanvv - HOSTVN.NET Technical</strong></p>
<p align="center"><strong>Home Page: <a href="https://hostvn.vn/">Hostvn Scripts Home page</a> , <a href="https://hostvn.net">Hostvn.net - Tên miền, Web Hosting, Email, VPS &amp; Dịch vụ lưu trữ Website</a></strong></p>
<p align="center"><strong>Document: https://help.hostvn.vn/</strong></p>
<p align="center"><strong>Groups Support: https://www.facebook.com/groups/hostvn.vn</strong></p>

<p align="center"> <img src="https://blog.hostvn.net/wp-content/uploads/2020/07/logo-big-2.png" /> </p>

########################################################################################

Script được viết bằng shell dùng để cài đặt LEMP Stack (Nginx - MariaDB - PHP-FPM) trên Ubuntu (18.04, 20.04), Debian 10.

<b>Vui lòng không sao chép, phát hành lại với mục đích thương mại, donate. Xin cảm ơn.</b>

## 1. Chi tiết Script:

### 1.1. Cài đặt

- Cập nhật liên tục, cung cấp Menu giúp dễ dàng thao tác, tự động cài đặt.
- Cài đặt các phần mềm với phiên bản mới nhất từ chính trang chủ.
- Mariadb: 10.5.
- Phiên bản Nginx Stable.
- Cho phép lựa chọn phiên bản PHP: 5.6, 7.0, 7.1, 7.2, 7.3, 7.4, 8.0
- phpMyAdmin 4.9.x nếu cài đặt PHP mặc định là 5.6, phpMyAdmin 5.0.x nếu cài đặt PHP mặc định là 7.x.
- Cấu hình Nginx FastCGI cache
- Cấu hình Nginx Pagespeed
- Cài đặt PHPMemcachedAdmin, phpRedisAdmin, Opcache Dashboard.
- Cài đặt memcached, redis cache. (Mặc định sẽ không bật).
- Cài đặt Fail2ban.
- Tích hợp Let's Encrypt SSL.
- Tích hợp CloudFlare DNS API giúp xác thực SSL nhanh hơn.
- Cung cấp Menu nâng cấp Nginx, MariaDB, PHP, Redis, Memcached, phpMyAdmin.
- Cài đặt WP-CLI, Composer, supervisor, Rclone.
- Cài đặt: ClamAV, ImunifyAV.
- KHÔNG THU THẬP BẤT KỲ THÔNG TIN GÌ TRÊN VPS CỦA BẠN.

### 1.2. Tối ưu

- Cấu hình tối ưu MySQL, Nginx, PHP, Opcache, Memcached, Redis phù hợp theo thông số của VPS.
- Cấu hình Brotli Compress.
- Cấu hình rewrite URL với một số mã nguồn phổ biến: WordPress, Laravel, Opencart, Magento, Drupal ...
- Cho phép chạy song song hai phiên bản PHP.
- Quản lý FTP dễ dàng trong trường hợp bạn thuê coder và chỉ muốn họ được phép access vào thư mục nhất định.
- Cho phép lựa chọn Redis Cache hoặc Memcached giúp tăng tốc độ website.
- Cho phép cấu hình tự động gia hạn Let's Encrypt.
- Menu hỗ trợ cấu hình SSL trả phí.
- Menu xem error log Nginx, Mysql, PHP và xem theo từng tên miền cụ thể.
- Cronjob tự động cập nhật dải IP mới nhất của Cloudflare cho Nginx và CSF Firewall.
- And More ...

### 1.3. Bảo mật

- Cấu hình tăng cường bảo mật từ tầng webserver.
- Cấu hình website chạy với user khác nhau hạn chế lây lan mã độc giữa các website.
- Disable các hàm nguy hiểm, bật open_basedir và một số config khác nhằm tăng cường bảo mật.
- Auto block bruteforce SSH, SFTP, Admin Tool .... với Fail2ban.
- Block run shell trong thư mục uploads WordPress. Chặn truy cập các thư mục, file nhạy cảm trên WordPress.
- Chống Bruteforce wp-admin.
- BLock, Unblock IP dễ dàng với menu quản lý Firewall.
- Disable User API - /wp-json/wp/v2/users - trên WordPress tránh lộ thông tin User.
- Thay đổi Port SSH tránh scan SSH.
- Cho phép thay đổi port Admin.
- Cho phép thay đổi port SSH/SFTP.
- Tự động tạo mật khẩu mạnh.
- Scan Malware với Clamav, ImunifyAV.
- And More ...

### 1.4. Quản lý WordPress

- Kiểm tra phiên bản WordPress đang sử dụng.
- Cập nhật WordPress.
- Cập nhật plugins.
- Tối ưu Database.
- Repair Database.
- Sao lưu dữ liệu (Local/GG Drive).
- Khôi phục dữ liệu (Local/GG Drive).
- Thay đổi tên miền.
- Đổi mật khẩu admin (wp-admin).
- Cài đặt WordPress tự động.
- Bật/tắt Yoast Seo config.
- Bật/Tắt Rank Math Seo config.
- Cấu hình Nginx với một số plugins cache phổ biến: WP-Rocket, w3 total cache, wp supercache, Cache Enabler, Swift Performance, Fast cache.
- Thêm cache key tránh trùng lặp nội dung giữa các site khi dùng memcached hoặc redis.
- Bật/tắt Debug.
- Bật/tắt chế độ bảo trì.
- Bật/Tắt disable xmlrpc (Mặc định sẽ disable xmlrpc để tránh bị khai thác DDOS).
- Huỷ kích hoạt toàn bộ plugins.
- Random database prefix khi sử dụng chức năng cài đặt tự động.
- And more ...

### 1.5. Sao lưu/Khôi phục dữ liệu
- Backup và restore dữ liệu từ Google Drive với Rclone.
- Backup, restore tại Local.
- Thiết lập số bản backup lưu trữ.
- Quản lý các bản sao lưu.
- Cho phép kết nối nhiều tài khoản Google Drive.

## 2. Yêu cầu

- VPS tối thiểu 512MB ram và chưa cài đặt bất kỳ dịch vụ nào.
- Ubuntu 18.04, 20.04, Debian 10.

## 3. Cách cài đặt

`curl -sO https://scripts.hostvn.net/install && chmod +x install && ./install`

## 4. Các tính năng sẽ phát triển

- Các tính năng theo request của người dùng

## 5. Cách sử dụng

- Vui lòng truy cập: https://help.hostvn.vn/

## 6. Nguồn download các phần mềm

- Nginx: http://nginx.org/en/download.html
- MariaDB: https://downloads.mariadb.org/
- PHP: https://www.php.net/
- phpMyAdmin: https://www.phpmyadmin.net/
- PHPMemcachedAdmin: https://github.com/elijaa/phpmemcachedadmin
- phpRedisAdmin: https://github.com/erikdubbelboer/phpRedisAdmin
- Rclone: https://rclone.org/
- WP-CLI: https://wp-cli.org/
- Composer: https://getcomposer.org/
- ClamAV: https://www.clamav.net/
- ImunifyAV: https://www.imunify360.com/antivirus

## 7. Contact

- Homepage: https://hostvn.vn , https://hostvn.net
- Group: hhttps://www.facebook.com/groups/hostvn.vn
- Email: Sanvv@hostvn.com

## 8. Feedback

- Do chưa có nhiều kinh nghiệm nên Scripts khó tránh khỏi thiếu sót, rất mong nhận được ý kiến đóng góp của các bạn để Scripts ngày càng hoàn thiện hơn.
- Mọi ý kiến đóng góp các bạn vui lòng gửi về sanvv@hostvn.com, Groups Facebook: https://www.facebook.com/groups/hostvn.vn hoặc tạo Github Issues.

## 9. Contributors & Credits
### Developers / Maintainers
- Sanvv

### Contributors



changelog.txt
v0.2.1

- Added: Multiple SSL server



rebuild_nginx.sh
#!/bin/bash

######################################################################
#           Auto Install & Optimize LEMP Stack on Ubuntu             #
#                                                                    #
#                Author: Sanvv - HOSTVN Technical                    #
#                  Website: https://hostvn.vn                        #
#                                                                    #
#              Please do not remove copyright. Thank!                #
#  Please do not copy under any circumstance for commercial reason!  #
######################################################################

apt-get install build-essential zlib1g-dev libpcre3 libpcre3-dev unzip uuid-dev -y

UPDATE_LINK="https://scripts.hostvn.net/ubuntu/update"
MODULE_PATH="/usr/share/nginx_module"
mkdir -p "${MODULE_PATH}"
NGINX_VERSION=$(curl -s ${UPDATE_LINK}/version | grep "nginx_version=" | cut -f2 -d'=')
NPS_VERSION=$(curl -s ${UPDATE_LINK}/version | grep "nps_version=" | cut -f2 -d'=')
ngx_cache_purge_version=$(curl -s ${UPDATE_LINK}/version | grep "ngx_cache_purge_version=" | cut -f2 -d'=')
more_clear_headers_v=$(curl -s ${UPDATE_LINK}/version | grep "more_clear_headers_v=" | cut -f2 -d'=')
openssl_version=$(curl -s ${UPDATE_LINK}/version | grep "openssl_version=" | cut -f2 -d'=')
pcre_version=$(curl -s ${UPDATE_LINK}/version | grep "pcre_version=" | cut -f2 -d'=')
zlib_version=$(curl -s ${UPDATE_LINK}/version | grep "zlib_version=" | cut -f2 -d'=')

cd "${MODULE_PATH}" || exit

rm -rf "${MODULE_PATH:?}"/*

wget -O- https://github.com/apache/incubator-pagespeed-ngx/archive/v"${NPS_VERSION}".tar.gz | tar -xz
nps_dir=$(find . -name "*pagespeed-ngx-${NPS_VERSION}" -type d)
cd "$nps_dir" || exit
NPS_RELEASE_NUMBER=${NPS_VERSION/beta/}
NPS_RELEASE_NUMBER=${NPS_VERSION/stable/}

psol_url=https://dl.google.com/dl/page-speed/psol/${NPS_RELEASE_NUMBER}.tar.gz
[ -e scripts/format_binary_url.sh ] && psol_url=$(scripts/format_binary_url.sh PSOL_BINARY_URL)
wget -O- "${psol_url}" | tar -xz

cd "${MODULE_PATH}" || exit
wget -O- http://nginx.org/download/nginx-"${NGINX_VERSION}".tar.gz | tar -xz
wget -O- http://scripts.hostvn.net/ubuntu/modules/ngx_cache_purge-"${ngx_cache_purge_version}".tar.gz | tar -xz
wget -O- http://scripts.hostvn.net/ubuntu/modules/openssl-OpenSSL_"${openssl_version}".tar.gz | tar -xz
wget -O- ftp://ftp.pcre.org/pub/pcre/pcre-"${pcre_version}".tar.gz | tar -xz
wget -O- https://www.zlib.net/zlib-"${zlib_version}".tar.gz | tar -xz
wget -O- http://scripts.hostvn.net/ubuntu/modules/headers-more-nginx-module-"${more_clear_headers_v}".tar.gz | tar -xz

git clone --depth 1 https://github.com/google/ngx_brotli
cd ngx_brotli && git submodule update --init

cd "${MODULE_PATH}"/nginx-"${NGINX_VERSION}"/ || exit
./configure \
    "--user=nginx" \
    "--group=nginx" \
    "--prefix=/usr" \
    "--sbin-path=/usr/sbin" \
    "--conf-path=/etc/nginx/nginx.conf" \
    "--pid-path=/var/run/nginx.pid" \
    "--http-log-path=/var/log/nginx/access_log" \
    "--error-log-path=/var/log/nginx/error_log" \
    "--without-mail_imap_module" \
    "--without-mail_smtp_module" \
    "--with-http_ssl_module" \
    "--with-http_realip_module" \
    "--with-http_stub_status_module" \
    "--with-http_gzip_static_module" \
    "--with-http_dav_module" \
    "--with-http_v2_module" \
    "--with-pcre=../pcre-${pcre_version}" \
    "--with-pcre-jit" \
    "--with-zlib=../zlib-${zlib_version}" \
    "--with-openssl=../openssl-OpenSSL_${openssl_version}" \
    "--with-openssl-opt=no-nextprotoneg" \
    "--add-module=../ngx_cache_purge-${ngx_cache_purge_version}" \
    "--add-module=../incubator-pagespeed-ngx-${NPS_VERSION}" \
    "--add-module=../headers-more-nginx-module-${more_clear_headers_v}" \
    "--add-module=../ngx_brotli" \
    "--with-cc-opt='-D FD_SETSIZE=32768'"

make && make install


menu/controller/cache/script/install_php_redis.sh
#!/bin/bash

######################################################################
#           Auto Install & Optimize LEMP Stack on Ubuntu             #
#                                                                    #
#                Author: Sanvv - HOSTVN Technical                    #
#                  Website: https://hostvn.vn                        #
#                                                                    #
#              Please do not remove copyright. Thank!                #
#  Please do not copy under any circumstance for commercial reason!  #
######################################################################

source /var/hostvn/menu/helpers/variable_common
source /var/hostvn/menu/helpers/function
DIR="/opt/php-extension"
MODULE_LINK="${HOMEPAGE_LINK}/modules"
IG_BINARY_VERSION=$(curl -s "${UPDATE_LINK}"/version | grep "igbinary_version=" | cut -f2 -d'=')
PHP_REDIS_VERSION=$(curl -s "${UPDATE_LINK}"/version | grep "php_redis_version=" | cut -f2 -d'=')
PHP1_VERSION=$(grep -w "php1_version" "/var/hostvn/.hostvn.conf" | cut -f2 -d'=')
PHP2_RELEASE=$(grep -w "php2_release" "/var/hostvn/.hostvn.conf" | cut -f2 -d'=')
PHP2_VERSION=$(grep -w "php2_version" "/var/hostvn/.hostvn.conf" | cut -f2 -d'=')
PHP1_INI_PATH="/etc/php/${PHP1_VERSION}/fpm/conf.d"
PHP2_INI_PATH="/etc/php/${PHP2_VERSION}/fpm/conf.d"
PHP1_CLI_PATH="/etc/php/${PHP1_VERSION}/cli/conf.d"
PHP2_CLI_PATH="/etc/php/${PHP2_VERSION}/cli/conf.d"

_install_ig_binary_php1(){
    if [[ "${PHP1_VERSION}" == "5.6" ]]; then
        IG_BINARY_VERSION="2.0.8"
    fi

    _cd_dir "${DIR}"
    if [ -f "igbinary-${IG_BINARY_VERSION}.tgz" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}".tgz
    fi
    if [ -d "igbinary-${IG_BINARY_VERSION}" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}"
    fi
    wget "${MODULE_LINK}"/igbinary-"${IG_BINARY_VERSION}".tgz
    tar -xvf igbinary-"${IG_BINARY_VERSION}".tgz
    _cd_dir igbinary-"${IG_BINARY_VERSION}"
    /usr/bin/phpize"${PHP1_VERSION}" && ./configure --with-php-config=/usr/bin/php-config"${PHP1_VERSION}"
    make && make install
    _cd_dir "${DIR}"
    rm -rf igbinary-"${IG_BINARY_VERSION}" igbinary-"${IG_BINARY_VERSION}".tgz

    rm -rf "$PHP1_INI_PATH"/30-igbinary.ini
    cat >> "$PHP1_INI_PATH/30-igbinary.ini" << EOF
extension=igbinary.so
EOF

    rm -rf "$PHP1_CLI_PATH"/30-igbinary.ini
    cat >> "$PHP1_CLI_PATH/30-igbinary.ini" << EOF
extension=igbinary.so
EOF
    systemctl restart php"${PHP1_VERSION}"-fpm
}

_install_php1_redis(){
    if [[ "${PHP1_VERSION}" == "5.6" ]]; then
        PHP_REDIS_VERSION="4.3.0"
    fi

    _cd_dir "${DIR}"
    if [ -f "redis-${PHP_REDIS_VERSION}.tgz" ]; then
        rm -rf redis-"${PHP_REDIS_VERSION}".tgz
    fi
    if [ -d "memcached-${PHP_REDIS_VERSION}" ]; then
        rm -rf redis-"${PHP_REDIS_VERSION}"
    fi
    wget "${MODULE_LINK}"/redis-"${PHP_REDIS_VERSION}".tgz
    tar -xvf redis-"${PHP_REDIS_VERSION}".tgz
    _cd_dir "${DIR}/redis-${PHP_REDIS_VERSION}"
    /usr/bin/phpize"${PHP1_VERSION}" && ./configure --enable-redis-igbinary --with-php-config=/usr/bin/php-config"${PHP1_VERSION}"
    make && make install
    cd "${DIR}" && rm -rf redis-"${PHP_REDIS_VERSION}".tgz redis-"${PHP_REDIS_VERSION}"

    rm -rf "${PHP1_INI_PATH}"/50-redis.ini
    cat >> "${PHP1_INI_PATH}/50-redis.ini" << EOF
extension=redis.so
EOF

    rm -rf "${PHP1_CLI_PATH}"/50-redis.ini
    cat >> "${PHP1_CLI_PATH}/50-redis.ini" << EOF
extension=redis.so
EOF
}

_install_ig_binary_php2(){
    if [[ "${PHP2_VERSION}" == "5.6" ]]; then
        IG_BINARY_VERSION="2.0.8"
    fi

    _cd_dir "${DIR}"
    if [ -f "igbinary-${IG_BINARY_VERSION}.tgz" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}".tgz
    fi
    if [ -d "igbinary-${IG_BINARY_VERSION}" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}"
    fi
    wget "${MODULE_LINK}"/igbinary-"${IG_BINARY_VERSION}".tgz
    tar -xvf igbinary-"${IG_BINARY_VERSION}".tgz
    _cd_dir "igbinary-${IG_BINARY_VERSION}"
    /usr/bin/phpize"${PHP2_VERSION}" && ./configure --with-php-config=/usr/bin/php-config"${PHP2_VERSION}"
    make && make install
    cd "${DIR}" && rm -rf igbinary-"${IG_BINARY_VERSION}" igbinary-"${IG_BINARY_VERSION}".tgz

    rm -rf "${PHP2_INI_PATH}"/30-igbinary.ini
    cat >> "${PHP2_INI_PATH}/30-igbinary.ini" << EOF
extension=igbinary.so
EOF

    rm -rf "$PHP2_CLI_PATH"/30-igbinary.ini
    cat >> "$PHP2_CLI_PATH/30-igbinary.ini" << EOF
extension=igbinary.so
EOF
    systemctl restart php"${PHP2_VERSION}"-fpm
}

_install_php2_redis(){
    if [[ "${PHP2_VERSION}" == "5.6" ]]; then
        PHP_REDIS_VERSION="4.3.0"
    fi

    _cd_dir "${DIR}"
    if [ -f "redis-${PHP_REDIS_VERSION}.tgz" ]; then
        rm -rf redis-"${PHP_REDIS_VERSION}".tgz
    fi
    if [ -d "redis-${PHP_REDIS_VERSION}" ]; then
        rm -rf redis-"${PHP_REDIS_VERSION}"
    fi
    wget "${MODULE_LINK}"/redis-"${PHP_REDIS_VERSION}".tgz
    tar -xvf redis-"${PHP_REDIS_VERSION}".tgz
    _cd_dir "${DIR}/redis-${PHP_REDIS_VERSION}"
    /usr/bin/phpize"${PHP2_VERSION}" && ./configure --enable-redis-igbinary --with-php-config=/usr/bin/php-config"${PHP2_VERSION}"
    make && make install
    cd "${DIR}" && rm -rf redis-"${PHP_REDIS_VERSION}".tgz redis-"${PHP_REDIS_VERSION}"

    rm -rf "${PHP2_INI_PATH}"/50-redis.ini
    cat >> "${PHP2_INI_PATH}/50-redis.ini" << EOF
extension=redis.so
EOF

    rm -rf "${PHP2_CLI_PATH}"/50-redis.ini
    cat >> "${PHP2_CLI_PATH}/50-redis.ini" << EOF
extension=redis.so
EOF
}

mkdir -p "${DIR}"
check_ig_binary_php1=$(php"${PHP1_VERSION}" -m | grep igbinary)
check_redis_php1=$(php"${PHP1_VERSION}" -m | grep redis)

if [ -z "$check_ig_binary_php1" ]; then
    _install_ig_binary_php1
    check_ig_binary_php1=$(php"${PHP1_VERSION}" -m | grep igbinary)
fi

if [[ -n "$check_ig_binary_php1" && -z "$check_redis_php1" ]]; then
    _install_php1_redis
fi

if [ "${PHP2_RELEASE}" == "yes" ]; then
    check_ig_binary_php2=$(php"${PHP2_VERSION}" -m | grep igbinary)
    check_redis_php2=$(php"${PHP2_VERSION}" -m | grep redis)

    if [ -z "$check_ig_binary_php2" ]; then
        _install_ig_binary_php2
        check_ig_binary_php2=$(php"${PHP2_VERSION}" -m | grep igbinary)
    fi

    if [[ -n "$check_ig_binary_php2" && -z "$check_redis_php2" ]]; then
        _install_php2_redis
    fi
fi

rm -rf "${DIR}"



menu/controller/cache/script/install_php_memcached.sh
#!/bin/bash

######################################################################
#           Auto Install & Optimize LEMP Stack on Ubuntu             #
#                                                                    #
#                Author: Sanvv - HOSTVN Technical                    #
#                  Website: https://hostvn.vn                        #
#                                                                    #
#              Please do not remove copyright. Thank!                #
#  Please do not copy under any circumstance for commercial reason!  #
######################################################################

source /var/hostvn/menu/helpers/variable_common
source /var/hostvn/menu/helpers/function
DIR="/opt/php-extension"
MODULE_LINK="${HOMEPAGE_LINK}/modules"
IG_BINARY_VERSION=$(curl -s "${UPDATE_LINK}"/version | grep "igbinary_version=" | cut -f2 -d'=')
PHP_MEMCACHED_VERSION=$(curl -s "${UPDATE_LINK}"/version | grep "php_memcached_version=" | cut -f2 -d'=')
PHP1_VERSION=$(grep -w "php1_version" "/var/hostvn/.hostvn.conf" | cut -f2 -d'=')
PHP2_RELEASE=$(grep -w "php2_release" "/var/hostvn/.hostvn.conf" | cut -f2 -d'=')
PHP2_VERSION=$(grep -w "php2_version" "/var/hostvn/.hostvn.conf" | cut -f2 -d'=')
PHP1_INI_PATH="/etc/php/${PHP1_VERSION}/fpm/conf.d"
PHP2_INI_PATH="/etc/php/${PHP2_VERSION}/fpm/conf.d"
PHP1_CLI_PATH="/etc/php/${PHP1_VERSION}/cli/conf.d"
PHP2_CLI_PATH="/etc/php/${PHP2_VERSION}/cli/conf.d"

_install_ig_binary_php1(){
    if [[ "${PHP1_VERSION}" == "5.6" ]]; then
        IG_BINARY_VERSION="2.0.8"
    fi

    _cd_dir "${DIR}"
    if [ -f "igbinary-${IG_BINARY_VERSION}.tgz" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}".tgz
    fi
    if [ -d "igbinary-${IG_BINARY_VERSION}" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}"
    fi
    wget "${MODULE_LINK}"/igbinary-"${IG_BINARY_VERSION}".tgz
    tar -xvf igbinary-"${IG_BINARY_VERSION}".tgz
    _cd_dir igbinary-"${IG_BINARY_VERSION}"
    /usr/bin/phpize"${PHP1_VERSION}" && ./configure --with-php-config=/usr/bin/php-config"${PHP1_VERSION}"
    make && make install
    _cd_dir "${DIR}"
    rm -rf igbinary-"${IG_BINARY_VERSION}" igbinary-"${IG_BINARY_VERSION}".tgz

    rm -rf "$PHP1_INI_PATH"/30-igbinary.ini
    cat >> "$PHP1_INI_PATH/30-igbinary.ini" << EOF
extension=igbinary.so
EOF

    rm -rf "$PHP1_CLI_PATH"/30-igbinary.ini
    cat >> "$PHP1_CLI_PATH/30-igbinary.ini" << EOF
extension=igbinary.so
EOF
    systemctl restart php"${PHP1_VERSION}"-fpm
}

_install_php1_memcached(){
    if [[ "${PHP1_VERSION}" == "5.6" ]]; then
        PHP_MEMCACHED_VERSION="2.2.0"
    fi

    _cd_dir "${DIR}"
    if [ -f "memcached-${PHP_MEMCACHED_VERSION}.tgz" ]; then
        rm -rf memcached-"${PHP_MEMCACHED_VERSION}".tgz
    fi
    if [ -d "memcached-${PHP_MEMCACHED_VERSION}" ]; then
        rm -rf memcached-"${PHP_MEMCACHED_VERSION}"
    fi
    wget "${MODULE_LINK}"/memcached-"${PHP_MEMCACHED_VERSION}".tgz
    tar -xvf memcached-"${PHP_MEMCACHED_VERSION}".tgz
    _cd_dir "memcached-${PHP_MEMCACHED_VERSION}"
    /usr/bin/phpize"${PHP1_VERSION}" && ./configure --enable-memcached-igbinary --with-php-config=/usr/bin/php-config"${PHP1_VERSION}"
    make && make install
    cd "${DIR}" && rm -rf memcached-"${PHP_MEMCACHED_VERSION}".tgz memcached-"${PHP_MEMCACHED_VERSION}"
    rm -rf "$PHP1_INI_PATH"/40-memcached.ini
    cat >> "$PHP1_INI_PATH/40-memcached.ini" << EOF
extension=memcached.so
EOF

    rm -rf "$PHP1_CLI_PATH"/40-memcached.ini
    cat >> "$PHP1_CLI_PATH/40-memcached.ini" << EOF
extension=memcached.so
EOF
    systemctl restart php"${PHP1_VERSION}"-fpm
}

_install_ig_binary_php2(){
    if [[ "${PHP2_VERSION}" == "5.6" ]]; then
        IG_BINARY_VERSION="2.0.8"
    fi

    _cd_dir "${DIR}"
    if [ -f "igbinary-${IG_BINARY_VERSION}.tgz" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}".tgz
    fi
    if [ -d "igbinary-${IG_BINARY_VERSION}" ]; then
        rm -rf igbinary-"${IG_BINARY_VERSION}"
    fi
    wget "${MODULE_LINK}"/igbinary-"${IG_BINARY_VERSION}".tgz
    tar -xvf igbinary-"${IG_BINARY_VERSION}".tgz
    _cd_dir "igbinary-${IG_BINARY_VERSION}"
    /usr/bin/phpize"${PHP2_VERSION}" && ./configure --with-php-config=/usr/bin/php-config"${PHP2_VERSION}"
    make && make install
    cd "${DIR}" && rm -rf igbinary-"${IG_BINARY_VERSION}" igbinary-"${IG_BINARY_VERSION}".tgz

    rm -rf "${PHP2_INI_PATH}"/30-igbinary.ini
    cat >> "${PHP2_INI_PATH}/30-igbinary.ini" << EOF
extension=igbinary.so
EOF

    rm -rf "$PHP2_CLI_PATH"/30-igbinary.ini
    cat >> "$PHP2_CLI_PATH/30-igbinary.ini" << EOF
extension=igbinary.so
EOF
    systemctl restart php"${PHP2_VERSION}"-fpm
}

_install_php2_memcached(){
    if [[ "${PHP2_VERSION}" == "5.6" ]]; then
        PHP_MEMCACHED_VERSION="2.2.0"
    fi

    _cd_dir "${DIR}"
    if [ -f "memcached-${PHP_MEMCACHED_VERSION}.tgz" ]; then
        rm -rf memcached-"${PHP_MEMCACHED_VERSION}".tgz
    fi
    if [ -d "memcached-${PHP_MEMCACHED_VERSION}" ]; then
        rm -rf memcached-"${PHP_MEMCACHED_VERSION}"
    fi
    wget "${MODULE_LINK}"/memcached-"${PHP_MEMCACHED_VERSION}".tgz
    tar -xvf memcached-"${PHP_MEMCACHED_VERSION}".tgz
    _cd_dir "memcached-${PHP_MEMCACHED_VERSION}"
    /usr/bin/phpize"${PHP2_VERSION}" && ./configure --enable-memcached-igbinary --with-php-config=/usr/bin/php-config"${PHP2_VERSION}"
    make && make install
    cd "${DIR}" && rm -rf memcached-"${PHP_MEMCACHED_VERSION}".tgz memcached-"${PHP_MEMCACHED_VERSION}"

    rm -rf "${PHP2_INI_PATH}"/40-memcached.ini
    cat >> "${PHP2_INI_PATH}/40-memcached.ini" << EOF
extension=memcached.so
EOF

    rm -rf "$PHP2_CLI_PATH"/40-memcached.ini
    cat >> "$PHP2_CLI_PATH/40-memcached.ini" << EOF
extension=memcached.so
EOF
    systemctl restart php"${PHP2_VERSION}"-fpm
}

mkdir -p "${DIR}"
check_ig_binary_php1=$(php"${PHP1_VERSION}" -m | grep igbinary)
check_memcached_php1=$(php"${PHP1_VERSION}" -m | grep memcached)

if [ -z "$check_ig_binary_php1" ]; then
    _install_ig_binary_php1
    check_ig_binary_php1=$(php"${PHP1_VERSION}" -m | grep igbinary)
fi

if [[ -n "$check_ig_binary_php1" && -z "$check_memcached_php1" ]]; then
    _install_php1_memcached
fi

if [ "${PHP2_RELEASE}" == "yes" ]; then
    check_ig_binary_php2=$(php"${PHP2_VERSION}" -m | grep igbinary)
    check_memcached_php2=$(php"${PHP2_VERSION}" -m | grep memcached)

    if [ -z "$check_ig_binary_php2" ]; then
        _install_ig_binary_php2
        check_ig_binary_php2=$(php"${PHP2_VERSION}" -m | grep igbinary)
    fi

    if [[ -n "$check_ig_binary_php2" && -z "$check_memcached_php2" ]]; then
        _install_php2_memcached
    fi
fi

cd /
rm -rf "${DIR}"
